Create a package that can pull singularity images to a permanent cache.

rhpvorderman opened the pull request:
This should initialize the singularity-permanent-cache package. After this we can test it a bit in production. If it works it can be published.

rhpvorderman committed:
add package skeleton

rhpvorderman committed:
create command line tool

rhpvorderman committed:
add a function for a default cache dir

rhpvorderman committed:
determine cache from environment

rhpvorderman committed:
remove impossible test

rhpvorderman committed:
add singularity command and options to replace uri special characters

rhpvorderman committed:
add pull image to cache command

rhpvorderman committed:
add simple file lock

rhpvorderman committed:
add tests for lock

rhpvorderman committed:
working test

rhpvorderman committed:
make sure test works correctly with multiprocessing.Lock()

rhpvorderman committed:
continue on testing locks

rhpvorderman committed:
add working test

rhpvorderman committed:
working test

rhpvorderman committed:
add logging

rhpvorderman committed:
use floats for faster testing

rhpvorderman committed:
allow testing with tox

rhpvorderman committed:
fix linting errors

rhpvorderman committed:
Add test for uri conversion

rhpvorderman committed:
Expand pull image to cache function. Add log messages

rhpvorderman committed:
complete main function

rhpvorderman committed:
fix linting issues

rhpvorderman committed:
Add some code documentation

rhpvorderman committed:
Make ui functional

rhpvorderman committed:
use filelock

rhpvorderman committed:
fix logging

rhpvorderman committed:
simpler lock test

rhpvorderman committed:
start on main testing

rhpvorderman committed:
move filelock test

rhpvorderman committed:
add tests for pull image to cache

rhpvorderman committed:
add travis matrix

rhpvorderman committed:
update version of singularity

rhpvorderman committed:
get singularity from neurodebian

rhpvorderman committed:
show singularity version

rhpvorderman committed:
install singularity from source

rhpvorderman committed:
fix matrix

rhpvorderman committed:
ubuntu packages

rhpvorderman committed:
different golang package

rhpvorderman committed:
update package cache

rhpvorderman committed:
update package cache

rhpvorderman committed:
Fix os.remove for 3.5

rhpvorderman committed:
proper __init__ with __all__

rhpvorderman committed:
add comment

rhpvorderman committed:
Write documentation

rhpvorderman committed:
add file lock to the documentation

rhpvorderman committed:
add codecov service yml

rhpvorderman committed:
remove redundant bandit dependency

vsoch commented on src/singularity_permanent_cache/singularity_permanent_cache.py:
```diff
@@ -0,0 +1,183 @@
+                    "Creating cache dir: {0}".format(str(cache)))
+        cache.mkdir(parents=True)
+
+    image_path = Path(cache, uri_to_filename(uri)).with_suffix(".sif")
```
Just to check - this cache is using an image named by the uri, and not the hash, so when used in place of normal singularity pulling we have a (sort) of double namespace, where some images are named by hash, others are named by uri?

rhpvorderman replied:
Yes. This is necessary because a hash lookup will connect to the internet. That means that when quay.io is down for example, your cache will not work anymore. The translation of uri -> image is one-to-one. Same uri, same image. Implementing multiple URIs for the same image requires more complex code, and is more prone to breaking due to internet dependencies.

Also, singularity's builtin cache already handles this use case. This cache was made for a use cache that the builtin cache does not address.

vsoch replied:
okay, makes sense.

vsoch commented on src/singularity_permanent_cache/singularity_permanent_cache.py:
```diff
@@ -0,0 +1,183 @@
+    lockfile_path = Path(cache, ".lock")
+
+    if not image_path.exists():
+        with SimpleUnixFileLock(str(lockfile_path)):
```
Wouldn't we want a name that ends with something with .tmpxxx in case we need to clean up? Or are you using the filename to honor the uri? Ideally we would want to have the container name present to mean that it's completely been pulled, and isn't a partial file so there isn't a race condition. My concern is having a pull started, something goes wrong, and an incomplete image (that looks like a .sif file) OR another node seeing that the image path exists and assuming it's good to use. Could you walk me through how this would work with multiple nodes running at the same time without having a race condition or using an incomplete file?

rhpvorderman replied:
>  Could you walk me through how this would work with multiple nodes running at the same time without having a race condition or using an incomplete file?

This script handles concurrency well when running *on one node only*. It was made to function with cromwell, which runs the job submission script on one node as well. 

> My concern is having a pull started, something goes wrong, and an incomplete image (that looks like a .sif file) OR another node seeing that the image path exists and assuming it's good to use. Could you walk me through how this would work with multiple nodes running at the same time without having a race condition or using an incomplete file?

You are right. I wonder what the `singularity pull` command does. I will get back on this.

rhpvorderman replied:
The issue should be addressed now
1. The filelock is now applied also to image existence checking. This means that the existence of an image can only be checked after another process is finished. If this other process happens to be pulling the image, this means the image will be completely pulled before any existence checks are run.
2. The image is now pulled to `***.sif.tmp` and once the pull is complete it will be moved to `***.sif`. If the singularity pull process exits with errors then the `***.sif` file will never be created.

Thanks for these great suggestions! 

As regards to the 'multiple nodes' issue. This is simply not possible without using some sort of server infrastructure. A filelock works reliably for processes on one node only. Multiple node locking is much harder, and can presumably not be solved with a simple script.

Single-node locking is sufficient for Cromwell's use case and I am a lazy programmer :wink:.

rhpvorderman replied:
I have added a note to the README as well explaining how the lock only works for multiple processes on the same node, to avoid any confusion. Thanks for pointing this out @vsoch!

vsoch replied:
Sure thing! I point out the multinode issue because we had a user in Europe that, by way of using a singularity exec over a pull first to use the same image binary, what it resulted in was thousands of nodes pulling the same 550 GB container at the same time, and not keeping any cache, so over a few days before my billing alerts went off for Singularity Hub there was substantial damage. So I bring it up only because I have real experience to suggest that not considering some multi-node usage case could have dire consequences. In that this repository isn't explicitly said to be used for cromwell, I think it would at least be worth adding a note about proper usage in the case of multiple nodes with different levels of support for cache, e.g.,:

 - if you have isolated, ephemeral nodes, you must pull once before, and upload the container to the node as a binary executable (the case I mentioned above) as the cache will never be honored.
 - if you have isolated, non-ephemeral nodes, you can likely use the exec as you normally would, each will download the image and then save to the cache. It would still be recommended to pull first on a single node and then distribute the container to the isolated nodes to avoid the first group of ~1000 pulls (depending on the size of the cluster)
 - if you have non-isolated, non-ephemeral nodes, you can also pull first, but actually your use case is best fitting to the library here. You can use it to generate an image in the cache, and then it will be found across nodes by way of the shared filesystem.

You can tweak those as you see fit! Then at least if something bad happens you can point to that section and say "Hey, I gave you a heads up about that!"

vsoch commented on src/singularity_permanent_cache/singularity_permanent_cache.py:
```diff
@@ -0,0 +1,183 @@
+    log_level = max(logging.WARNING + (args.quiet - args.verbose) * 10, 0)
+    log = logging.getLogger()  # gets the root logger.
+    logging.basicConfig()  # This adds the default handler to the root logger.
+    log.setLevel(log_level)
```
Could we have an argument that would let us just see settings / config (e.g., the cache that would be used?)

rhpvorderman replied:
I will check if this can easily be implemented.

rhpvorderman replied:
Well, apparently it can. I added a `--which-cache`.

rhpvorderman committed:
Always use lock before checking if image exists

rhpvorderman committed:
Pull to temporary image first

rhpvorderman committed:
Which cache action added.

rhpvorderman committed:
Added a note on the filelock

rhpvorderman committed:
prevent errors by using exist_ok

rhpvorderman committed:
Do not use with_suffix for the image name
This is necessary because with_suffix replaces the original suffix.
For example: docker_quay.io_biocontainers_fastqc_0.11.9 -> docker_quay.io_biocontainers_fastqc_0.11.sif is wrong.
This should be docker_quay.io_biocontainers_fastqc_0.11.9.sif.

rhpvorderman committed:
update development status

rhpvorderman committed:
add setup configuration to include license file

vsoch commented on README.rst:
```diff
@@ -0,0 +1,97 @@
+                            'docker://debian:buster-slim'
+
+    optional arguments:
+      -h, --help            show this help message and exit
```
Could we update here to add `--which-cache` and encourage the user to check it before running blindly?

rhpvorderman replied:
The readme is updated with the which cache.

As regards to encouraging the users: by default `SINGULARITY_CACHEDIR` is not set, even if Singularity is installed. `SINGULARITY_PERMANENTCACHEDIR` is specific to this program and is also not set by default. If these two variables are not set, and the program is run withour a `--cache-dir` argument, the program will crash. In case the user has set `SINGULARITY_CACHEDIR` before because they wanted that cache on a specific filesystem, will also have the permanent cache there. This will happen without prompt, but the program will emit a warning that the permanent cache dir is not yet created before creating it. In this particular case I assume that the reasons for having the singularity cache in this particular place also apply to the permanent cache. The chance of people running this program and having the cache end up somewhere where they do not want it is very slim IMO.

rhpvorderman committed:
add which cache to readme

vsoch reviewed (approved):
Just tested locally - it works as expected, and I like it a lot! I think this is a great start and good for merge. Does merging into develop imply that it won't be released yet?

rhpvorderman commented:
@vsoch, since I expect the syntax not to be changed I think I can aim for a 1.0.0 version, however I feel that still more testing is needed (I also tested on our own cluster and fixed a few oversights).

What I can do now, is release a 1.0.0alpha version.

This whay people can easily install the package via PYPI, (or download the script from the github page) and provide feedback. This way we can capture some corner-case bugs before we go to the 1.0.0beta and 1.0.0 proper releases.
